src/node.h

```diff
@@ -378,6 +378,84 @@ inline void NODE_SET_PROTOTYPE_METHOD(v8::Local<v8::FunctionTemplate> recv,
+template <class T>
+class WeakData {
+ public:
+  WeakData(v8::Isolate* isolate, v8::Local<v8::Value> destination) {
+    persistent.Reset(isolate, destination);
+    persistent.SetWeak(this, DeleteMe, v8::WeakCallbackType::kParameter);
+  }
+
+  virtual ~WeakData() {
+    if (!persistent.IsEmpty()) {
+      assert(persistent.IsNearDeath());
+      persistent.ClearWeak();
+      persistent.Reset();
+    }
+  }
+
+  static T* FromInfo(const v8::FunctionCallbackInfo<v8::Value>& info) {
+    return reinterpret_cast<T*>(info.Data().As<v8::External>()->Value());
+  }
+ private:
+  v8::Persistent<v8::Value> persistent;
+
+  static void DeleteMe(const v8::WeakCallbackInfo<WeakData>& info) {
+    delete info.GetParameter();
+  }
+};
+
+inline void SetMethodWithData(v8::Local<v8::Context> context,
+                              v8::Local<v8::Object> recv,
+                              const char* name,
+                              v8::FunctionCallback callback,
+                              void* data) {
+  v8::Isolate* isolate = context->GetIsolate();
+  v8::HandleScope handle_scope(isolate);
+  v8::Local<v8::FunctionTemplate> t =
+      v8::FunctionTemplate::New(isolate,
+                                callback,
+                                v8::External::New(isolate, data));
+  v8::MaybeLocal<v8::Function> maybe_fn = t->GetFunction(context);
+  assert(!maybe_fn.IsEmpty() && "Function must not be empty");
+  v8::Local<v8::Function> fn = maybe_fn.ToLocalChecked();
+
+  v8::MaybeLocal<v8::String> maybe_fn_name =
+      v8::String::NewFromUtf8(isolate, name, v8::NewStringType::kNormal);
+  assert(!maybe_fn_name.IsEmpty() && "Function name must not be empty");
+  v8::Local<v8::String> fn_name = maybe_fn_name.ToLocalChecked();
+
+  fn->SetName(fn_name);
+
+  v8::Maybe<bool> result = recv->Set(context, fn_name, fn);
+  assert(result.FromMaybe(false) && "Property assignment must succeed");
+}
+
+inline void SetPrototypeMethodWithData(v8::Local<v8::Context> context,
+                                       v8::Local<v8::FunctionTemplate> recv,
+                                       const char* name,
+                                       v8::FunctionCallback callback,
+                                       void* data) {
+  v8::Isolate* isolate = context->GetIsolate();
+  v8::HandleScope handle_scope(isolate);
+  v8::Local<v8::Signature> s = v8::Signature::New(isolate, recv);
+  v8::Local<v8::FunctionTemplate> t =
+      v8::FunctionTemplate::New(isolate,
+                                callback,
+                                v8::External::New(isolate, data),
+                                s);
+
+  v8::MaybeLocal<v8::String> maybe_fn_name =
+      v8::String::NewFromUtf8(isolate, name, v8::NewStringType::kNormal);
+  assert(!maybe_fn_name.IsEmpty() && "Function name must not be empty");
+  v8::Local<v8::String> fn_name = maybe_fn_name.ToLocalChecked();
+
+  t->SetClassName(fn_name);
+
+  recv->PrototypeTemplate()->Set(fn_name, t);
+}
+
```

Comment:
I wouldn’t be a fan of including this in our public API, unless there’s a reason why we would export it from core

We originally had more core of this sort around (e.g. `node_object_wrap.h`), but it’s a good idea that this generally moved into Nan territory? I’d suggest PR’ing it there